### ISSUE TITLE
Go command API restructuring

### ISSUE THREAD

**Issue by MiguelPires** [zeebe-io/zeebe#3741]
**Description**

Currently, each command in Go API is composed of several interfaces that force the user to provide the required information before sending the command or providing non-essential information. In my opinion, there a few disadvantages to this approach but the one I'd like to focus on relates to backwards compatibility. Whenever we make any change regarding the command API, we break compatibility (even if we don't consider breakage caused by struct embedding as in the [Go1 compat guarantee](https://golang.org/doc/go1compat#expectations)). This happens even when adding a field to a command request (we need to change the command's interfaces) or when adding a new command (we change the Client interface). I think it'd be better to return structs instead of interfaces since it'd make these changes non-breaking (again, ignoring struct embedding). Having an interface is good because it allows the user to decouple himself from our implementation and more easily implement a mock of Zeebe. However, since Go has implicit interfaces it's still possible to take advantage of that without us having to provide it. The user can just define an interface that Zeebe matches and it's implicitly implemented (simple example of this: https://play.golang.org/p/62bCWdK9kOy).

Other disadvantages of the current design:

* idiomatic for Java, not Go
* all of the interfaces make it very difficult to document the API (https://godoc.org/github.com/zeebe-io/zeebe/clients/go/pkg/commands#ActivateJobsCommandStep1 how can we document these interfaces in a way that's useful for someone reading the docs?)
* our command API is huge (some command have 3 interfaces and one struct - all exported) which might confuse the user when reading docs or even code

Reasons to keep it like it is:
* the way we chain the interfaces makes it easy for the user to know what piece of information he needs to provide next (this is a nice feature and makes it really easy to use it at first, especially if you use an IDE)
* this would be a pretty big breaking change obviously

Resources:
* https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8
* https://github.com/golang/go/wiki/CodeReviewComments#interfaces "Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values"

@menski @npepinpe

**Comment by npepinpe** [comment]
Good point; my (brief) experience in Go is that interfaces are indeed a way for a package to define its external dependencies, and not to return things to the outside. Since interfaces are implicitly implemented, it's more about defining a contract for incoming entities than it is about providing abstracted ones when returning anything.

**Comment by MiguelPires** [comment]
Another thing to be refactored: the job worker doesn't use the ActivateJobsCommand in `zbc` instead it reimplements the logic to poll jobs and insert them into a channel. Besides duplicating code it also creates some dependency issues. For instance, to provide retry/refresh capabilities we wanted to depend on the CredentialsProvider. In order to avoid a cyclic dependency between `zbc` and `worker`, this interface needs to be moved to `worker` even though it makes more sense to have in the client package. Ideally, the worker uses the command doesn't reimplement the retry logic, the credentials provider interface is used in a single place (`zbc`) and them the implementation could be isolated into its own package.

**Comment by MiguelPires** [comment]
After refactoring the client, we should also add example files to help our users understand how the client is supposed to be used. If we do them right they would be automatically embedded in the documentation which would greatly improve our Go docs. This also ensures that our documentation is up to date with our code, since the examples are actual code that has to compile
https://blog.golang.org/examples

**Comment by Zelldon** [comment]
I think this should be revisited by your team @camunda/zeebe-process-automation . Not sure whether it is even possible to change the go API now, if not then we can also just close the issue

**Comment by korthout** [comment]
@Zelldon I've moved the issue back into `Inbox`. We'll discuss it in the upcoming team meeting or planning.

**Comment by Zelldon** [comment]
Thanks nico 👍

**Comment by korthout** [comment]
The team discussed and sees no way to improve this in the near future. We should however look into this in the longer term. @felix-mueller Please keep this in mind for the product. Please also note that this issue may result in backwards incompatible changes.
